File: Lung/MasksToLobes/batch_generator.py
import numpy as np
import random
import ants
import antspynet

def batch_generator(batch_size=32,
                    segmentation_images=None,
                    priors=None,
                    do_histogram_intensity_warping=True,
                    do_simulate_bias_field=True,
                    do_add_noise=True,
                    do_data_augmentation=True,
                    use_two_outputs=False):

    if segmentation_images is None:
        raise ValueError("Input images must be specified.")

    while True:

        image_size = priors[0].shape
        number_of_channels = 1 + len(priors)

        X = np.zeros((batch_size, *image_size, number_of_channels))
        for b in range(batch_size):
            for p in range(len(priors)):
                X[b,:,:,:,p+1] = priors[p]

        Y  = np.zeros((batch_size, *image_size))

        batch_count = 0

        while batch_count < batch_size:
            i = random.sample(list(range(len(segmentation_images))), 1)[0]

            seg = ants.image_read(segmentation_images[i])
            h1 = ants.threshold_image(seg, 0, 0, 0, 1)

            if do_histogram_intensity_warping: # and random.sample((True, False), 1)[0]:
                break_points = [0.2, 0.4, 0.6, 0.8]
                displacements = list()
                for b in range(len(break_points)):
                    displacements.append(abs(random.gauss(0, 0.05)))
                    if random.sample((True, False), 1)[0]:
                        displacements[b] *= -1
                h1 = antspynet.histogram_warp_image_intensities(h1,
                    break_points=break_points, clamp_end_points=(True, False),
                    displacements=displacements)

            if do_add_noise: # and random.sample((True, False), 1)[0]:
                h1 = (h1 - h1.min()) / (h1.max() - h1.min())
                noise_parameters = (0.0, random.uniform(0, 0.05))
                h1 = ants.add_noise_to_image(h1, noise_model="additivegaussian", noise_parameters=noise_parameters)

            if do_simulate_bias_field: # and random.sample((True, False), 1)[0]:
                h1_field = antspynet.simulate_bias_field(h1, sd_bias_field=0.05) 
                h1 = ants.iMath(h1, "Normalize") * (h1_field + 1)
                h1 = (h1 - h1.min()) / (h1.max() - h1.min())

            # h1_pre = antspynet.preprocess_brain_image(h1, truncate_intensity=(0.01, 0.995), do_bias_correction=False, do_denoising=False, verbose=False)
            # h1 = h1_pre['preprocessed_image']
            # h1 = (h1 - h1.mean()) / h1.std()

            if do_data_augmentation == True:
                data_augmentation = antspynet.randomly_transform_image_data(h1,
                    [[h1]],
                    [seg],
                    number_of_simulations=1,
                    transform_type='deformation',
                    sd_affine=0.01,
                    deformation_transform_type="bspline",
                    number_of_random_points=1000,
                    sd_noise=2.0,
                    number_of_fitting_levels=4,
                    mesh_size=1,
                    sd_smoothing=4.0,
                    input_image_interpolator='linear',
                    segmentation_image_interpolator='nearestNeighbor')
                h1 = data_augmentation['simulated_images'][0][0]
                seg = data_augmentation['simulated_segmentation_images'][0]

            # ants.image_write(ants.from_numpy(h1_array), "h1_" + str(batch_count) + ".nii.gz")
            # ants.image_write(ants.from_numpy(seg_array), "seg_" + str(batch_count) + ".nii.gz")
            # print("batch_count = ", str(batch_count))

            X[batch_count,:,:,:,0] = h1.numpy()
            Y[batch_count,:,:,:] = np.round(seg.numpy())

            batch_count = batch_count + 1
            if batch_count >= batch_size:
                break

        labels = np.array((0, 1, 2, 3, 4, 5))

        encY = antspynet.encode_unet(Y.astype('int'), labels)
            
        if use_two_outputs:
            Y2 = np.sum(encY[:,:,:,:,1:], axis=4)
            yield X, [encY, Y2], [None, None]
        else:    
            yield X, encY, [None]

File: DeepAtropos/batchGenerator.R
batchGenerator <- function( batchSize = 32L,
                            imageSize = c( 64, 64, 64 ),
                            template = NULL,
                            images = NULL,
                            segmentationImages = NULL,
			    segmentationLabels = NULL,
                            doRandomContralateralFlips = TRUE,
                            doDataAugmentation = TRUE
 )
{                                     

  if( is.null( template ) )
    {
    stop( "No reference template specified." )
    }
  if( is.null( images ) )
    {
    stop( "Input images must be specified." )
    }
  if( is.null( segmentationImages ) )
    {
    stop( "Input masks must be specified." )
    }
  if( is.null( segmentationLabels ) )
    {
    stop( "segmentationLabels must be specified." )
    }

  currentPassCount <- 0L

  function()
    {
    # Shuffle the data after each complete pass

    if( ( currentPassCount + batchSize ) >= length( images ) )
      {
      # shuffle the source data
      sampleIndices <- sample( length( images ) )
      images <- images[sampleIndices]
      segmentationImages <- segmentationImages[sampleIndices]

      currentPassCount <- 0L
      }

    batchIndices <- currentPassCount + 1L:batchSize

    batchImages <- images[batchIndices]
    batchSegmentationImages <- segmentationImages[batchIndices]

    X <- array( data = 0, dim = c( batchSize, imageSize, 1 ) )    
    Y <- array( data = 0L, dim = c( batchSize, imageSize ) )

    currentPassCount <<- currentPassCount + batchSize

    # pb <- txtProgressBar( min = 0, max = batchSize, style = 3 )
    
    i <- 1
    while( i <= batchSize )
      {
      # setTxtProgressBar( pb, i )

      image <- NULL 
      batchImage <- antsImageRead( batchImages[i] )
      batchSegmentationImage <- antsImageRead( batchSegmentationImages[i] )
      if( doRandomContralateralFlips && sample( c( TRUE, FALSE ) ) )
        {
        image <- reflectImage( batchImage, axis = 0, verbose = FALSE )  
        segmentationImage <- reflectImage( batchSegmentationImage, axis = 0, verbose = FALSE )
        } else {
        image <- batchImage  
        segmentationImage <- batchSegmentationImage
        }
      
      # centerOfMassTemplate <- getCenterOfMass( template )
      # centerOfMassImage <- getCenterOfMass( segmentationImage )
      # xfrm <- createAntsrTransform( type = "Euler3DTransform",
      #   center = centerOfMassTemplate,
      #   translation = centerOfMassImage - centerOfMassTemplate )
      # warpedImage <- applyAntsrTransformToImage( xfrm, image, template )
      # warpedMask <- thresholdImage( applyAntsrTransformToImage( xfrm, segmentationImage, template ), 0.5, 1.0, 1, 0 )

      warpedImage <- image
      warpedMask <- segmentationImage
      
      if( doDataAugmentation == TRUE )
        {
        dataAugmentation <- 
          randomlyTransformImageData( template, 
          list( list( warpedImage ) ),
          list( warpedMask ),
          numberOfSimulations = 1, 
          transformType = 'affineAndDeformation', 
          sdAffine = 0.01,
          deformationTransformType = "bspline",
          numberOfRandomPoints = 1000,
          sdNoise = 2.0,
          numberOfFittingLevels = 4,
          meshSize = 1,
          sdSmoothing = 4.0,
          inputImageInterpolator = 'linear',
          segmentationImageInterpolator = 'nearestNeighbor' )

        simulatedImage <- dataAugmentation$simulatedImages[[1]][[1]]
        simulatedArray <- as.array( simulatedImage )
        simulatedArray <- ( simulatedArray - mean( simulatedArray ) ) / sd( simulatedArray )

        simulatedMask <- dataAugmentation$simulatedSegmentationImages[[1]]
        sumulatedMaskArray <- as.array( simulatedMask )


        # antsImageWrite( simulatedImage, paste0( "./TempData/simBrainImage", i, ".nii.gz" ) )
        # antsImageWrite( simulatedSegmentationImage, paste0( "./TempData/simSegmentationImage", i, ".nii.gz" ) )

        X[i,,,,1] <- simulatedArray[13:172,15:206,1:160]
        Y[i,,,] <- simulatedMaskArray[13:172,15:206,1:160]

        i <- i + 1  
        } else {
        warpedArray <- as.array( warpedImage )
        warpedArray <- ( warpedArray - mean( warpedArray ) ) / sd( warpedArray )        
        warpedMaskArray <- as.array( warpedMask )

        # antsImageWrite( warpedImage, paste0( "./TempData/wBrainImage", i, ".nii.gz" ) )
        # antsImageWrite( warpedSegmentationImage, paste0( "./TempData/wSegmentationImage", i, ".nii.gz" ) )

        X[i,,,,1] <- warpedArray[13:172,15:206,1:160]
        Y[i,,,] <- warpedMaskArray[13:172,15:206,1:160]

        i <- i + 1  
        }
      }
    # stop( "Done testing.")  
    # cat( "\n" )

    encodedY <- encodeUnet( Y, segmentationLabels )

    return( list( X, encodedY ) )
    }
}
